📚 Add a tutorial for how to build a java destination (#4384)
airbytehq/airbyte: docs/contributing-to-airbyte/tutorials/building-a-java-destination.md

### Step 2: Build the newly generated destination

You can build the destination by running:

```bash
# Must be run from the Airbyte project root
./gradlew :airbyte-integrations:connectors:destination-<name>:build
```

this compiles the java code for your destination and builds a Docker image with the connector. At this point, we haven't implemented anything of value yet, but once we do, you'll use this command to compile your code and Docker image. 

{% hint style="info" %}
    Airbyte uses Gradle to manage Java dependencies. To add dependencies for your connector, manage them in the `build.gradle` file inside your connector's directory.   
{% endhint %}

#### Iterating on your implementation

We recommend the following ways of iterating on your connector as you're making changes:

* Test-driven development (TDD) in Java 
* Test-driven development (TDD) using Airbyte's Acceptance Tests
* Directly running the docker image 

> **Test-driven development in Java**

Reviewer: I feel like you may as well make these H4 instead of just bolding them since you did that with the previous subheading. Same for the following bolded subheaders.
```suggestion
#### Test-driven development in Java
```

$ ./generate.sh
```

Select the `Java Destination` template and then input the name of your connector. We'll refer to the destination as `<name>-destination` in this tutorial, but you should replace `<name>` with the actual name you used for your connector e.g: `BigQueryDestination` or `bigquery-destination`. 

### Step 2: Build the newly generated destination

You can build the destination by running:

```bash
# Must be run from the Airbyte project root
./gradlew :airbyte-integrations:connectors:destination-<name>:build
```

this compiles the java code for your destination and builds a Docker image with the connector. At this point, we haven't implemented anything of value yet, but once we do, you'll use this command to compile your code and Docker image. 

{% hint style="info" %}
    Airbyte uses Gradle to manage Java dependencies. To add dependencies for your connector, manage them in the `build.gradle` file inside your connector's directory.   
{% endhint %}

#### Iterating on your implementation

We recommend the following ways of iterating on your connector as you're making changes:

> 1. Test-driven development in Java 

Author: done

2. Test-driven development using Airbyte's Acceptance Tests
3. Directly running the docker image 

**Test-driven development in Java**
This should feel like a standard flow for a Java developer: you make some code changes then run java tests against them. You can do this directly in your IDE, but you can also run all unit tests via Gradle by running the command to build the connector: 

```
./gradlew :airbyte-integrations:connectors:destination-<name>:build
```

This will build the code and run any unit tests. This approach is great when you are testing local behaviors and writing unit tests. 

> **TDD using acceptance tests & integration tests**

Reviewer 2: Nit. An acronym is better introduced by putting it in a pair of parentheses when the full term, test-driven development, is first mentioned in this doc:

> Test-driven development (TDD)

Reference: https://blog.apastyle.org/apastyle/abbreviations/#Q2

Author: done

> Airbyte provides a standard test suite (dubbed "Acceptance Tests") that is run against every destination. The objective of these tests is to provide some "free" baseline tests that ensure that the basic functionality of the destination works. One approach to developing your connector is to simply run the tests between each change and use the feedback from them to guide your development.

Reviewer 2: Nit. Some of the sentences can be more straightforward. For example:

```suggestion
Airbyte provides a standard test suite (dubbed "Acceptance Tests") that runs against every destination. They are "free" baseline tests to ensure the basic functionality of the destination. When developing a connector, you can simply run the tests between each change and use the feedback to guide your development.
```

Author: agreed, thanks for the suggestion

./gradlew :airbyte-integrations:connectors:destination-<name>:build

# Then use the following commands to run it
# Runs the "spec" command, used to find out what configurations are needed to run a connector 
docker run --rm airbyte/destination-<name>:dev spec

# Runs the "check" command, used to validate if the input configurations are valid
docker run --rm -v $(pwd)/secrets:/secrets airbyte/destination-<name>:dev check --config /secrets/config.json

# Runs the "write" command which reads records from stdin and writes them to the underlying destination
docker run --rm -v $(pwd)/secrets:/secrets -v $(pwd)/sample_files:/sample_files airbyte/destination-<name>:dev write --config /secrets/config.json --catalog /sample_files/configured_catalog.json
```

Note: Each time you make a change to your implementation you need to re-build the connector image via `./gradlew :airbyte-integrations:connectors:destination-<name>:build`.

The nice thing about this approach is that you are running your destination exactly as it will be run by Airbyte. The tradeoff is that iteration is slightly slower, because you need to re-build the connector between each change.

### Step 3: Implement `spec`

> Each destination contains a specification that describes what inputs it needs in order to run. Defining the specification is a good place to start when developing your destination. The easiest way is to use [JsonSchema](https://json-schema.org/) to define what the inputs are \(e.g. username and password\). Here's [an example](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json) of what the `spec.json` looks like for the postgres destination.

Reviewer 2: It feels useful to keep the reference to JsonSchema in case people need to check the syntax. Maybe something like this:

```suggestion
Each destination contains a specification written in JsonSchema that describes its inputs. Defining the specification is a good place to start when developing your destination. Check out the documentation [here](https://json-schema.org/) to learn the syntax. Here's [an example](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json) of what the `spec.json` looks like for the postgres destination.
```

Your generated template should have the spec file in `airbyte-integrations/connectors/destination-<name>/src/main/resources/spec.json`. Edit it and you should be done with this step. 

For more details on what the spec is, you can read about the Airbyte Protocol [here](../../understanding-airbyte/airbyte-specification.md).

> {% hint style="info" %}

Author: done

While developing, we recommend storing any credentials in `secrets/config.json`. Any `secrets` directory in the Airbyte repo is gitignored by default.

Implement the `check` method in the generated file `<Name>Destination.java`. Here's an [example implementation](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java#L94) from the BigQuery destination.  

Verify that the method is working by placing your config in `secrets/config.json` then running: 

```
# First build the connector
./gradlew :airbyte-integrations:connectors:destination-<name>:build

# Run the check method
docker run -v $(pwd)/secrets:/secrets --rm airbyte/destination-<name>:dev check --config /secrets/config.json
```

> ### Step 5: Implement `write`

Author: fixed in https://github.com/airbytehq/airbyte/pull/4418

The `write` operation is the main workhorse of a destination connector: it reads input data from the source and writes it to the underlying destination. It takes as input the config file used to run the connector as well as the configured catalog: the file used to describe the schema of the incoming data and how it should be written to the destination. 

To implement the `write` Airbyte operation, implement the `getConsumer` method in your generated `<Name>Destination.java` file. Here are some example implementations from different destination conectors:
 

> * [BigQuery](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java#L188) 

Reviewer 3: Thinking out loud: is it worth leaving an info section here about the Abstract Destination concept for destinations supported by JDBC?

Author: BQ is not JDBC - i added PG as an example of JDBC

To setup acceptance Tests for your connector, follow the `TODO`s in the generated file `<name>DestinationAcceptanceTest.java`. Once setup, you can run the tests using `./gradlew :airbyte-integrations:connectors:destination-<name>:integrationTest`. Make sure to run this command from the Airbyte repository root.

### Step 7: Write unit tests and/or integration tests
The Acceptance Tests are meant to cover the basic functionality of a destination. Think of it as the bare minimum required for us to add a destination to Airbyte. You should probably add some unit testing or custom integration testing in case you need to test additional functionality of your destination.

#### Step 8: Update the docs

Each connector has its own documentation page. By convention, that page should have the following path: in `docs/integrations/destinations/<destination-name>.md`. For the documentation to get packaged with the docs, make sure to add a link to it in `docs/SUMMARY.md`. You can pattern match doing that from existing connectors.

> 

Reviewer 2: Nit. Three newlines here.
